Add possibility to add bcs on edges/vertices

lijas opened the pull request:
Useful for for example "simply supported" bcs. 
Also useful for shell elements which only has edges.

See Cantilever.ipynb for example usage

KristofferC commented on src/Dofs/ConstraintHandler.jl:
```diff
@@ -145,7 +157,7 @@ function _add!(ch::ConstraintHandler, dbc::Dirichlet, bcfaces::Set{Tuple{Int,Int
     constrained_dofs = Int[]
     #_celldofs = fill(0, ndofs_per_cell(ch.dh))
-    for (cellidx, faceidx) in bcfaces
+    for  (cellidx, faceidx) in bcfaces
```
```suggestion
    for (cellidx, faceidx) in bcfaces
```

KristofferC commented on docs/src/literate/incompressible_elasticity.jl:
```diff
@@ -157,7 +157,7 @@ function assemble_up!(Ke, fe, cell, cellvalues_u, cellvalues_p, facevalues_u, gr
     ## is in our `"traction"` faceset.
     @inbounds for face in 1:nfaces(cell)
-        if onboundary(cell, face) && (cellid(cell), face) ∈ getfaceset(grid, "traction")
+        if onboundary(cell, face) && FaceIndex(cellid(cell), face) ∈ getfaceset(grid, "traction")
```
Hm, I guess this is pretty breaking. Perhaps we can leave `FaceIndex ` as `const FaceIndex = Tuple{Int, Int}` but use struct for the others?

KristofferC commented on examples/cantilever.ipynb:
```diff
@@ -244,42 +212,18 @@
-     "output_type": "stream",
-     "text": [
-      " \u001b[1m------------------------------------------------------------------\u001b[22m\n",
-      "                           Time                   Allocations      \n",
```
Did this table go away?

KristofferC commented on src/Dofs/ConstraintHandler.jl:
```diff
@@ -229,6 +241,7 @@ function _update!(values::Vector{Float64}, f::Function, faces::Set{Tuple{Int,Int
     _celldofs = fill(0, ndofs_per_cell(dh, _tmp_cellid))
 
     for (cellidx, faceidx) in faces
+
```
```suggestion
```

KristofferC commented on src/Dofs/ConstraintHandler.jl:
```diff
@@ -395,22 +409,30 @@ end
 #Function for adding constraint when using multiple celltypes
 function add!(ch::ConstraintHandler, fh::FieldHandler, dbc::Dirichlet)
-    dbc.faces isa Set{Tuple{Int}} && error("Can only apply boundary conditions to faces (not nodes).")
+    
```
```suggestion
```

lijas commented:
I added a simple example with shell elements, which is the strongest motivation why one needs to be able to add BCs on edges/vertices. I am planning to write some text for the example so that it can be added to the other tutorials. 

I would  also like some feedback on the structure of the code. Currently in the ` add!`-function for the `ConstraintHandler`, I changed it so that it creates `BCValues` for either faces, edges or vertices:

```
function add!(ch::ConstraintHandler, dbc::Dirichlet)
...
functype = getgeometryfunction(eltype(dbc.bcset)) # eltype(dbc.bcset) is either FaceIndex, EdgeIndex or VertexIndes
bcvalue = BCValues(interpolation, JuAFEM.default_interpolation(celltype), functype)
_add!(ch, dbc, dbc.bcset, interpolation, field_dim, field_offset(ch.dh, dbc.field_name), bcvalue, functype)
```

where

```
getgeometryfunction(::Type{FaceIndex}) = JuAFEM.faces
getgeometryfunction(::Type{EdgeIndex}) = JuAFEM.edges
getgeometryfunction(::Type{VertexIndex}) = JuAFEM.vertices
```

Do you think it is clear what  `getgeometryfunction` is doing? Is there a better name? Or should we skip it all together and just do

```
if eltype(dbc.bcset) isa FaceIndex
    functype = JuAFEM.faces
elseif eltype(dbc.bcset) isa EdgeIndex
   functype = JuAFEM.edges
elseif eltype(dbc.bcset) isa VertexIndex
   functype = JuAFEM.vertices
end
bcvalue = BCValues(interpolation, JuAFEM.default_interpolation(celltype), functype)
_add!(ch, dbc, dbc.faces, interpolation, field_dim, field_offset(ch.dh, dbc.field_name), bcvalue, functype)
```

Any feedback would be appriciated..

KristofferC commented:
I think doing the explicit `if` makes sense for now. If the same type of code occurs in other places, it could be changed out to dispatch.

lijas commented:
Close due to #343
